src/Microsoft.Framework.Runtime.Sources/Impl/ApplicationGlobalData.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Framework.Runtime.Sources.Impl
+{
+    // Helper class to completely centralize all of the application global data logic and avoid #if statements elsewhere.
+    internal class ApplicationGlobalData
+    {
+        private readonly IApplicationEnvironment _hostEnvironment;
+#if DNXCORE50
+        private readonly object _lock;
+        private readonly Dictionary<string, object> _appGlobalData;
```

Comment:
Seems legit to me. The goal here is to create an application-global data store, not to provide access to AppDomain.CurrentDomain.Get/SetData.

@davidfowl ?